### PR TITLE
Breaking changes (erdjs 10): improve contract interactions and interpretation of contract results

### PR DESCRIPTION
**Pull request by andreibancioiu**
Related docs PR: https://github.com/ElrondNetwork/docs.elrond.com/pull/318

**Breaking changes**

 - `ExecutionResultsBundle` and `QueryResponseBundle` have been removed, and replaced by `TypedOutcomeBundle` (and its untyped counterpart, `UntypedOutcomeBundle`).
 - `SmartContractResults` has been changed to not use the concepts `immediate result` and `resulting calls` anymore. Instead, interpreting `SmartContractResults.items` is now the responsibility of the `ResultsParser` (on which the contract controllers depend).
 - Redesigned `QueryResponse`, changed most of its public interface. Results interpretation is now the responsibility of the results parser, called by the smart contract controllers.
 - `interpretExecutionResults()` and `interpretExecutionResults()` do not exist on the `Interaction` object anymore. Now, querying / executing an interaction against the controller will return the interpreted results.
 - `TokenIdentifierValue` is constructed using a `string`, not a `buffer`. It's `valueOf()` it now a string, as well.
 - The `Interaction` constructor does not receive the `interpretingFunction` parameter anymore.
 - `Interaction.getInterpretingFunction()` and `Interaction.getExecutingFunction()` have been removed, replaced by `Interaction.getFunction()`.
 - `DefaultInteractionRunner` has been removed, and replaced by **smart contract controllers**.
 - `StrictChecker` has been renamed to `InteractionChecker`. It's public interface - the function `checkInteraction()` - has changed as well (it also requires the endpoint definition now, as a second parameter).
 - The functions `getReceipt()`, `getSmartContractResults()` and `getLogs()` of `TransactionOnNetwork` have been removed. The underlying properties are now public.
 - Renamed `OptionValue.newMissingType()` to `OptionValue.newMissingTyped()`
 - Queries with a return code different than `Ok` do not automatically throw an exception anymore (`assertSuccess()` has to be called explicitly in order to throw).
 
**Other changes**

 - `SmartContract`, in addition to `methods`, now also has a `methodAuto` object that allows one to create interactions without explicitly specifying the types of the arguments. Automatic type inference (within erdjs' typesystem) is leveraged. The type inference system was implemented in the past, in the `nativeSerializer` component - PR https://github.com/ElrondNetwork/elrond-sdk-erdjs/pull/9 by @claudiu725.
 - Added utility function `getFieldValue()` on `Struct` and `EnumValue`.
 - Refactoring in the `networkProvider` package (under development, in order to merge the provider interfaces under a single one)
 - Added utility function `Interaction.useThenIncrementNonceOf()`
 - Fixed `nativeSerializer` to not depend on `SmartContract`, `ContractWrapper` and `TestWallet` anymore (gathered under an interface).
 - Replaced the old `lottery-egld` with the new `lottery-esdt` in integration tests.
 - Added missing tests for some components: `nativeSerializer`, `struct`, `enum`.
 - Added utility function `OptionalValue.newMissing()`. Added "fake" covariance wrt. "null type parameter" (when value is missing) on `OptionalType`.
 - Added utility functions (typed value factories): `createListOfAddresses`, `createListOfTokensIdentifiers`.

### CODE REVIEW

**Review by claudiu725 — CHANGES_REQUESTED** [review]

File: src/smartcontracts/index.ts [hunk 1]
```diff
@@ -8,17 +8,19 @@ export * from "./argSerializer";
...
 export * from "./interface";
 export * from "./nativeSerializer";
 export * from "./query";
 export * from "./queryResponse";
+export * from "./resultsParser";
 export * from "./returnCode";
 export * from "./smartContract";
 export * from "./smartContractResults";
-export * from "./strictChecker";
+export * from "./interactionChecker";
```
Comment (camilbancioiu) [inline src/smartcontracts/index.ts:23]: Duplicate import? See line 14.

**Review by camilbancioiu — APPROVED** [review]

File: src/smartcontracts/typesystem/factory.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import { List } from "./generic";
+import { TokenIdentifierValue } from "./tokenIdentifier";
+
+export function createListOfAddresses(addresses: Address[]): List {
+    let addressesTyped = addresses.map(address => new AddressValue(address));
+    let list = List.fromItems(addressesTyped);
+    return list;
+}
+
+export function createListOfTokensIdentifiers(identifiers: string[]): List {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply src/smartcontracts/typesystem/factory.ts:12]: Fixed.

File: src/smartcontracts/index.ts (hunk 1 above)
Comment (andreibancioiu) [reply src/smartcontracts/index.ts:23]: Fixed. Yes, it was duplicated by mistake :)

File: src/smartcontracts/smartContractController.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        this.provider = provider;
+    }
+
+    async deploy(transaction: Transaction): Promise<{ transactionOnNetwork: TransactionOnNetwork, bundle: UntypedOutcomeBundle }> {
+        Logger.info(`SmartContractController.deploy [begin]: transaction = ${transaction.getHash()}`);
+
+        await transaction.send(this.provider);
+        await transaction.awaitExecuted(this.provider);
+        let transactionOnNetwork = await transaction.getAsOnNetwork(this.provider);
+        let bundle = this.parser.parseUntypedOutcome(transactionOnNetwork);
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply src/smartcontracts/smartContractController.ts:47]: Currently, deployments aren't performed via Interaction objects. Perhaps we should define a Deployment object, as well. Do you think we can postpone this to another PR? :pray:

[the remark this answers is not in the document]
Comment (claudiu725) [reply src/smartcontracts/smartContractController.ts:47]: Sure, we can postpone :+1:

**Review by claudiu725 — APPROVED** [review]

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@elrondnetwork/erdjs",
-  "version": "9.2.3",
+  "version": "10.0.0-beta.1",
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply package.json:3]: Indeed, we have that:

https://github.com/ElrondNetwork/elrond-sdk-erdjs/blob/better-interactions/.github/workflows/erdjs-publish-alpha-beta.yml#L29

**Review by camilbancioiu — APPROVED** [review]

File: src/smartcontracts/smartContract.ts [hunk 5]
```diff
@@ -36,6 +38,15 @@ export class SmartContract implements ISmartContract {
...
     public readonly methods: { [key: string]: (args?: TypedValue[]) => Interaction } = {};
 
+    /**
+     * This object contains a function for each endpoint defined by the contract.
+     * (a bit similar to web3js's "contract.methods").
+     * 
+     * This is an alternative to {@link methods}. 
+     * Unlike {@link methods}, automatic type inference (wrt. ABI) is applied when using {@link methodsAuto}.
+     */
+    public readonly methodsAuto: { [key: string]: (args?: any[]) => Interaction } = {};
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply src/smartcontracts/smartContract.ts:48]: Will be done in next PR.

**Pull request merged**